package org.apache.commons.cli.help;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Stream;
import org.apache.commons.cli.DeprecatedAttributes;
import org.apache.commons.cli.Option;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class OptionFormatterTest_testAsSyntaxOption {

    public static Stream<Arguments> deprecatedAttributesData() {
        final List<Arguments> lst = new ArrayList<>();
        final DeprecatedAttributes.Builder daBuilder = DeprecatedAttributes.builder();
        lst.add(Arguments.of(daBuilder.get(), "[Deprecated]"));
        daBuilder.setSince("now");
        lst.add(Arguments.of(daBuilder.get(), "[Deprecated since now]"));
        daBuilder.setForRemoval(true);
        lst.add(Arguments.of(daBuilder.get(), "[Deprecated for removal since now]"));
        daBuilder.setSince(null);
        lst.add(Arguments.of(daBuilder.get(), "[Deprecated for removal]"));
        daBuilder.setForRemoval(false);
        daBuilder.setDescription("Use something else");
        lst.add(Arguments.of(daBuilder.get(), "[Deprecated. Use something else]"));
        daBuilder.setForRemoval(true);
        lst.add(Arguments.of(daBuilder.get(), "[Deprecated for removal. Use something else]"));
        daBuilder.setForRemoval(false);
        daBuilder.setSince("then");
        lst.add(Arguments.of(daBuilder.get(), "[Deprecated since then. Use something else]"));
        daBuilder.setForRemoval(true);
        lst.add(Arguments.of(daBuilder.get(), "[Deprecated for removal since then. Use something else]"));
        return lst.stream();
    }

    private void assertEquivalent(final OptionFormatter formatter, final OptionFormatter formatter2) {
        assertEquals(formatter.toSyntaxOption(), formatter2.toSyntaxOption());
        assertEquals(formatter.toSyntaxOption(true), formatter2.toSyntaxOption(true));
        assertEquals(formatter.toSyntaxOption(false), formatter2.toSyntaxOption(false));
        assertEquals(formatter.getOpt(), formatter2.getOpt());
        assertEquals(formatter.getLongOpt(), formatter2.getLongOpt());
        assertEquals(formatter.getBothOpt(), formatter2.getBothOpt());
        assertEquals(formatter.getDescription(), formatter2.getDescription());
        assertEquals(formatter.getArgName(), formatter2.getArgName());
        assertEquals(formatter.toOptional("foo"), formatter2.toOptional("foo"));
    }

    @Test
    void testAsSyntaxOption() {
        OptionFormatter underTest;
        Option option = Option.builder().option("o").longOpt("opt").hasArg().get();
        underTest = OptionFormatter.from(option);
        assertEquals("[-o <arg>]", underTest.toSyntaxOption(), "optional arg failed");
        option = Option.builder().option("o").longOpt("opt").hasArg().argName("other").get();
        underTest = OptionFormatter.from(option);
        assertEquals("[-o <other>]", underTest.toSyntaxOption(), "optional 'other' arg failed");
        option = Option.builder().option("o").longOpt("opt").hasArg().required().argName("other").get();
        underTest = OptionFormatter.from(option);
        assertEquals("-o <other>", underTest.toSyntaxOption(), "required 'other' arg failed");
        option = Option.builder().option("o").longOpt("opt").required().argName("other").get();
        underTest = OptionFormatter.from(option);
        assertEquals("-o", underTest.toSyntaxOption(), "required no arg failed");
        option = Option.builder().option("o").argName("other").get();
        underTest = OptionFormatter.from(option);
        assertEquals("[-o]", underTest.toSyntaxOption(), "optional no arg arg failed");
        option = Option.builder().longOpt("opt").hasArg().argName("other").get();
        underTest = OptionFormatter.from(option);
        assertEquals("[--opt <other>]", underTest.toSyntaxOption(), "optional longOpt 'other' arg failed");
        option = Option.builder().longOpt("opt").required().hasArg().argName("other").get();
        underTest = OptionFormatter.from(option);
        assertEquals("--opt <other>", underTest.toSyntaxOption(), "required longOpt 'other' arg failed");
        option = Option.builder().option("ot").longOpt("opt").hasArg().get();
        underTest = OptionFormatter.from(option);
        assertEquals("[-ot <arg>]", underTest.toSyntaxOption(), "optional multi char opt arg failed");
    }
}
